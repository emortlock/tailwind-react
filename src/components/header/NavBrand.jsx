import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'

const NavBrand = ({ theme, header: { style }, is, children, ...rest }) => {
  const styleProps = {
    flex: [true, 'no-shrink'],
    items: 'center',
    h: 12,
    m: { r: theme.spacing.lg },
    text: style.text,
  }

  const ariaProps = !(typeof is === 'string' && is.startsWith('h'))
    ? {
        role: 'heading',
        'aria-level': 1,
      }
    : {}

  return (
    <BaseComponent is={is} noUnderline {...styleProps} {...ariaProps} {...rest}>
      {children}
    </BaseComponent>
  )
}

NavBrand.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    style: PropTypes.object.isRequired,
  }),
}

NavBrand.defaultProps = {
  is: 'div',
  children: undefined,
  header: {},
}

export { NavBrand as component }
export default withTheme(NavBrand)
