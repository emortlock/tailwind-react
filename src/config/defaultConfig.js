export default {
  radius: 'rounded',
  spacing: {
    zero: 0,
    sm: 2,
    md: 4,
    lg: 8,
  },
  container: {
    sm: 'xl',
    md: '3xl',
    lg: '4xl',
  },
  text: {
    size: {
      body: ['sm', 'base', 'lg'],
      title: ['lg', 'xl', '2xl', '3xl', '4xl', '5xl'],
    },
    family: {
      body: 'sans',
      subtitle: 'sans',
      title: 'sans',
    },
  },
  baseColors: {
    primaryLight: 'blue-light',
    primary: 'blue',
    primaryDark: 'blue-dark',
    secondaryLight: 'grey',
    secondary: 'grey-dark',
    secondaryDark: 'grey-darker',
    successLight: 'green-light',
    success: 'green',
    successDark: 'green-dark',
    dangerLight: 'red-light',
    danger: 'red',
    dangerDark: 'red-dark',
    warningLight: 'orange-lighter',
    warning: 'orange-light',
    warningDark: 'orange',
    infoLight: 'teal-lighter',
    info: 'teal-light',
    infoDark: 'teal',
  },
  textColors: {
    body: 'grey-darkest',
    link: 'blue-dark',
    linkDark: 'blue-darker',
    emphasis: 'black',
    on: {
      primaryLight: 'white',
      primary: 'white',
      primaryDark: 'white',
      secondaryLight: 'white',
      secondary: 'white',
      secondaryDark: 'white',
      successLight: 'white',
      success: 'white',
      successDark: 'white',
      dangerLight: 'white',
      danger: 'white',
      dangerDark: 'white',
      warningLight: 'black',
      warning: 'black',
      warningDark: 'black',
      infoLight: 'black',
      info: 'black',
      infoDark: 'black',
    },
  },
}
