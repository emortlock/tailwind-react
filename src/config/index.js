export { default as defaultConfig } from './defaultConfig'
export { default as TailwindConfig } from './TailwindConfig'
export { default as withConfig } from './withConfig'
