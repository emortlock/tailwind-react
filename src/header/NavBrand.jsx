import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import { withTailwind } from '../tailwind'
import { filterProps } from '../utils'

const NavBrand = ({ theme, is, children, className, ...rest }) => {
  const Component = is

  return (
    <Component
      {...filterProps(rest, ['header'])}
      className={classnames(
        `mr-${theme.spacing.lg}`,
        'flex items-center flex-no-shrink h-12',
        className,
      )}
    >
      {children}
    </Component>
  )
}

NavBrand.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  className: PropTypes.string,
}

NavBrand.defaultProps = {
  is: 'div',
  children: undefined,
  className: undefined,
}

export { NavBrand as component }
export default withTheme(withTailwind(NavBrand))
