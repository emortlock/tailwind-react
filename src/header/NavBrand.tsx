import type { BoxProps } from 'tailwind-react-primitives'
import { WithTheme } from '../theme'
import { WithHeader } from './Header'

import React from 'react'

import { RawBox as Box } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface NavBrandProps<E extends HTMLElement = HTMLDivElement>
  extends BoxProps<E>,
    WithTheme,
    WithHeader {}

export const RawNavBrand = ({
  theme,
  header: { style, screen } = { style: {}, screen: 'lg' },
  as = 'div',
  children,
  ...rest
}: NavBrandProps) => {
  const responsive = screen
    ? {
        [`m-${screen}`]: { r: theme.spacing.lg },
      }
    : {}

  const aria = !(typeof as === 'string' && as.startsWith('h'))
    ? {
        'role': 'heading',
        'aria-level': 1,
      }
    : {}

  return (
    <Box
      as={as}
      inlineBlock
      flex={[true, 'shrink-0']}
      items="center"
      h={12}
      text={style?.text || theme.textColors.on.primary}
      {...responsive}
      {...aria}
      {...rest}
    >
      {children}
    </Box>
  )
}

export const NavBrand = withTheme(RawNavBrand)
