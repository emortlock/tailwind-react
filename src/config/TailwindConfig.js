import React from 'react'
import defaultConfig from './defaultConfig'

export default React.createContext(defaultConfig)
