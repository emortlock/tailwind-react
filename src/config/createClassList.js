import merge from 'lodash.merge'
import defaultConfig from './defaultConfig'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectConfig = {}) => {
  const config = merge(defaultConfig, projectConfig)

  return [
    config.radius,
    ...reduceMultiplePrefixes(config.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(config.container, 'max-w'),
    ...mapArray(config.text.size.body, 'text'),
    ...mapArray(config.text.size.title, 'text'),
    ...mapObject(config.text.family, 'font'),
    ...reduceMultiplePrefixes(config.baseColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(config.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
