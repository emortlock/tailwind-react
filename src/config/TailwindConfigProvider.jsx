import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultConfig from './defaultConfig'
import TailwindConfig from './TailwindConfig'

const TailwindConfigProvider = ({ config, children }) => {
  const mergedConfig = merge(defaultConfig, config)

  return (
    <TailwindConfig.Provider value={mergedConfig}>
      {children}
    </TailwindConfig.Provider>
  )
}

TailwindConfigProvider.propTypes = {
  config: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindConfigProvider.defaultProps = {
  config: {},
  children: undefined,
}

export default TailwindConfigProvider
