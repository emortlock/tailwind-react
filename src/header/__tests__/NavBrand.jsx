import React from 'react'
import { shallow } from 'enzyme'

import { component as NavBrand } from '../NavBrand'
import { defaultConfig } from '../../config'

const setup = (testProps = {}) => {
  const props = Object.assign({ config: defaultConfig }, testProps)

  const wrapper = shallow(<NavBrand {...props}>Tailwind React</NavBrand>)

  return {
    props,
    wrapper,
  }
}

describe('NavBrand', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
