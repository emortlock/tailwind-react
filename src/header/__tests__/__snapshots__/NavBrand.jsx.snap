// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBrand renders matching snapshot 1`] = `
<div
  className="mr-8 flex items-center flex-no-shrink h-12"
>
  Tailwind React
</div>
`;
